Faster B0 shimming

Aurelienpujolm opened the pull request:
## Checklist

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied the relevant labels to this PR
- [x] I've added relevant tests for my contribution
- [x] I've updated the documentation and/or added correct docstrings
- [x] I've assigned a reviewer

<!--- For the title, please observe the following rules:
	- Provide a concise and self-descriptive title
	- Do not include the applicable issue number in the title, do it in the PR body
	- If the PR is not ready for review, convert it to a draft.
-->

## Description
This PR will be trying to make the B0 shimming as fast as possible 
Specially, this PR will 

- Change the function to minimize to a newest one which is faster to minimize (residuals mse)
- Find a way to plot the newest currents coefficients as fast as possible 
- Reduced the processing time of the sequencer 

## Linked issues
This pull request fixes #412 about the time of the B0 shimming

Aurelienpujolm committed:
Add a faster solver.

Add a new class to allow the user to optimize the currents with scipy.optimize.minimize by using the residuals_mse function, and the jacobian of it

Aurelienpujolm committed:
Use the smaller dataset

Aurelienpujolm committed:
Faster the sequencer
1) Removed the multiprocessing, if the user is using a mac (Spawn method makes the optimization slower)
2) If there is no correction in a slice then it's possible to reduce the calculation time of _eval_static_shim and _cal_anat_orient. By not making all of the time consumming process.

Aurelienpujolm committed:
Make _plot_coef faster
Removes all of the useless subplot where there is no corrections.
Change the size of the figure to make it smaller if there is less subplot to show
Add a subplot to show an example of the current correction in an unshimmed slice
If the bounds are the same in every channel, then don't display it for all of them, but instead plot a single line to represent all of them

Aurelienpujolm committed:
Creating the Shimmed anat orient faster
-Reshape the coils_anat matrix into a 2D matrix
-Change the dimension of the correction matrix to a 2D one (This allow to make the calculation a lot faster, by removing a dimension for the np.sum)
- Make a list of the slices to shim to make the code faster and more readable

Aurelienpujolm committed:
Fixed a typo mistake in calc_shimmed_anat_orient

Aurelienpujolm committed:
Merge branch 'master' into ap/faster-solver

# Conflicts:
#	shimmingtoolbox/cli/b0shim.py
#	shimmingtoolbox/shim/sequencer.py

Aurelienpujolm committed:
Make sure that the merge with master is working
Use the new opt_criteria from master
In Eval_static_shim, only print in the debugging if there is a correction in a slice

Aurelienpujolm committed:
Make sure that the optimization is working with the merge
Inplement a jacobian when the function to minimize is mse
Delete lsq_faster_solver which is now obsolet

Aurelienpujolm committed:
Fix all of the problem with the merging
Removed all of the comment about least-squares-faster
Fixed some typo problem
Put opt_criteria, everytime that the optimize function is called

Aurelienpujolm committed:
Removed the jacobian in real time shimming

Aurelienpujolm committed:
Removed an useless test

po09i commented on shimmingtoolbox/cli/b0shim.py:
```diff
@@ -978,19 +979,31 @@ def _get_current_shim_settings(json_data):
                 pres_probe_max=None, units='', bounds=None):
-    n_shims = static_coefs.shape[0]
-    fig = Figure(figsize=(8, 4 * n_shims), tight_layout=True)
+    # We want to find how much slices are not shimmed to have a smaller file and reduced the plotting and saving time
```
```suggestion
    # We want to find which slices are not shimmed to have a smaller file size and reduce the plot saving time
```

po09i commented on shimmingtoolbox/cli/b0shim.py:
```diff
@@ -1018,32 +1031,59 @@ def _plot_coefs(coil, slices, static_coefs, path_output, coil_number, rt_coefs=N
+        _add_sub_figure(i_shim, fig, number_slices_shimmed, nb_slices, rt_coefs, pres_probe_min,
+                        pres_probe_max, bounds, min_y, max_y,
+                        units, static_coefs, i_slice)
+    # Print a sublplot for all the non shimmed slices
```
```suggestion
    # Print a subplot for all the non shimmed slices
```

po09i commented on shimmingtoolbox/cli/b0shim.py:
```diff
@@ -978,19 +979,31 @@ def _get_current_shim_settings(json_data):
+            # Get a string with the number of all the unshimmed slices
+            list_slice_wo_shimmed = list_slice_wo_shimmed + str(slices[i_shim]) + ","
+            # Get the last index where the shimmed correction is null will permit to plot the last one as an example
+            # for all the other one
```
```suggestion
            # Get the last index where the shimmed correction is null, this will allow to plot the last one as an example
            # for all the other ones
```

po09i commented on shimmingtoolbox/cli/b0shim.py:
```diff
@@ -1018,32 +1031,59 @@ def _plot_coefs(coil, slices, static_coefs, path_output, coil_number, rt_coefs=N
             max_y = np.array(static_coefs).max()
+    nb_slices = 0
+    for i_shim in shimmed_slice:
+        nb_slices = nb_slices + 1
```
```suggestion
    for nb_slices, i_shim in enumerate(shimmed_slice):
```

po09i replied:
The names of the variables are confusing.

po09i commented on shimmingtoolbox/cli/b0shim.py:
```diff
@@ -1018,32 +1031,59 @@ def _plot_coefs(coil, slices, static_coefs, path_output, coil_number, rt_coefs=N
+        if np.all(bounds[:, 0] == bounds[0, 0]) and np.all(bounds[:, 1] == bounds[0, 1]):
+            ax.hlines(bounds[0, 0], 0 - len_hline_bounds, n_channels + len_hline_bounds, colors='r',
+                      label='bounds',
+                      capstyle='projecting')
```
```suggestion
                      label='bounds', capstyle='projecting')
```

po09i commented on shimmingtoolbox/optimizer/lsq_optimizer.py:
```diff
@@ -1,11 +1,11 @@
 #!/usr/bin/python3
 # -*- coding: utf-8 -*-
+from timeit import timeit
```
```suggestion

```

po09i commented on shimmingtoolbox/optimizer/lsq_optimizer.py:
```diff
@@ -1,11 +1,11 @@
 import scipy.optimize as opt
 from typing import List
 import warnings
-
```
Add the space back

po09i replied:
I fixed it

po09i commented on shimmingtoolbox/optimizer/lsq_optimizer.py:
```diff
@@ -47,6 +47,7 @@ def __init__(self, coils: ListCoil, unshimmed, affine, opt_criteria='mse', reg_f
         }
         if opt_criteria in lsq_residual_dict:
             self._criteria_func = lsq_residual_dict[opt_criteria]
+            self.method_to_minimize = opt_criteria
```
```suggestion
            self.opt_criteria = opt_criteria
```

You should name the input the same as the attribute name so that it is less confusing

po09i commented on shimmingtoolbox/optimizer/lsq_optimizer.py:
```diff
@@ -148,13 +149,21 @@ def _apply_sum_constraint(inputs, indexes, coef_sum_max):
-                                   bounds=self.merged_bounds,
-                                   constraints=tuple(scipy_constraints),
-                                   options={'maxiter': 1000})
+        if self.method_to_minimize == 'mse':
```
```suggestion
        if self.opt_criteria == 'mse':
```

po09i commented on shimmingtoolbox/optimizer/lsq_optimizer.py:
```diff
@@ -208,6 +217,32 @@ def _initial_guess_zeros(self):
+        np.mean((unshimmed_vec + np.sum(coil_mat * coef, axis=1, keepdims=False)) ** 2) / factor+\
+           (self.reg_factor * np.mean(np.abs(coef) / self.reg_factor_channel))
+        If you want to see the mathematical expression for the jacobian, of this function or for the mae one,
+         please contact : aurelienpujolmpro@gmail.com
```
```suggestion
        please contact : aurelienpujolmpro@gmail.com
```

po09i commented on shimmingtoolbox/shim/sequencer.py:
```diff
@@ -176,55 +174,56 @@ def _eval_static_shim(opt: Optimizer, nii_fieldmap_orig, nii_mask, coef, slices,
-                     f"current: \n{coef[i_shim, :]}")
+        # Calculate shimmed values
+        if not np.any(coef[i_shim]):
+            shimmed[..., i_shim] = unshimmed
```
```suggestion
            shimmed[..., i_shim] = unshimmed
            continue
```

You can then unindent that whole else block

po09i commented on shimmingtoolbox/shim/sequencer.py:
```diff
@@ -281,12 +280,15 @@ def _cal_shimmed_anat_orient(coefs, coils, nii_mask_anat, nii_fieldmap, slices,
+    dimx = np.shape(coils_anat)[0]
+    dimy = np.shape(coils_anat)[1]
+    for i_shim in list_shim_slice:
+        # We want to do the np.sum with a 2D matrix to get it aster
```
```suggestion
        # We want to do the np.sum with a 2D matrix so that it is faster
```

po09i commented on shimmingtoolbox/shim/sequencer.py:
```diff
@@ -281,12 +280,15 @@ def _cal_shimmed_anat_orient(coefs, coils, nii_mask_anat, nii_fieldmap, slices,
+        # We want to do the np.sum with a 2D matrix to get it aster
+        coils_anat_reduced = np.reshape(coils_anat[:, :, slices[i_shim], :], (-1, nb_channel))
+        corr = np.sum(coefs[i_shim] * coils_anat_reduced, axis=1, keepdims=False)
+        shimmed_anat_orient[..., slices[i_shim]] += np.reshape(corr, (dimx, dimy, 1))
```
I think this can crash because the slice dimension is not always 1, I suggest using the way I had showed you.

po09i commented on shimmingtoolbox/shim/sequencer.py:
```diff
@@ -936,46 +938,55 @@ def select_optimizer(method, unshimmed, affine, coils: ListCoil, opt_criteria, p
 
 
-def _optimize(optimizer: Optimizer, nii_mask_anat, slices_anat, shimwise_bounds=None,
+@timeit
```
```suggestion
@timeit
```
Calculating the time on a "private" function is fine in the PR but not to be displayed to the user

po09i commented on test/cli/test_cli_b0shim.py:
```diff
@@ -81,18 +83,21 @@ def test_cli_dynamic_sph(self, nii_fmap, nii_anat, nii_mask, fm_data, anat_data)
                                              '--mask', fname_mask,
                                              '--scanner-coil-order', '2',
                                              '--regularization-factor', '0.1',
+                                             # '-v', 'debug',
```
```suggestion

```

po09i commented on test/cli/test_cli_b0shim.py:
```diff
@@ -81,18 +83,21 @@ def test_cli_dynamic_sph(self, nii_fmap, nii_anat, nii_mask, fm_data, anat_data)
             assert res.exit_code == 0
             assert os.path.isfile(os.path.join(tmp, "coefs_coil0_Prisma_fit_gradient_coil.txt"))
 
+   
```
```suggestion
```

Aurelienpujolm commented:
With all of the commits, the B0 shimming should now be way faster, and the shimming should be as efficient than before. As we can see in the pictures below :  
![image](https://user-images.githubusercontent.com/63396133/208184695-c0c241d1-3823-489a-a56d-4a3a92f04120.png)
_Figure 1 : Efficiency of the shimming in the ap/faster-solver branch_
![image](https://user-images.githubusercontent.com/63396133/208184671-8e23aee6-5e66-425a-aaed-2233ca5a7f45.png)
 _Figure 2 : Efficiency of the shimming in the master branch_
On a mac os computer (Dango), with a processor of 3.7 GHz, the processing time of a dynamic B0 shimming of the dataset test is now around 3 seconds (It was around 12 seconds before).
Also for a bigger dataset :  so with a coil of more channels ( 32 channels) , and a bigger anatomic image ([174,174,50] ), fieldmap([112,112,70]) and mask ([112,112,70]) , the processing time of a dynamic B0 shimming is now around 130 seconds vs 210 seconds before this PR ! (If you want to have this dataset to try it out, please contact [aurelienpujolm@outlook.fr](mailto:aurelienpujolm@outlook.fr))
However, some modifications can still be make, to improve the time,  such as :
- Find a way to fix the multiprocessing on a mac os, using a fast multiprocessing method should really help to make the optimization faster : 90 seconds without mp on the big dataset and 45 seconds with the multiprocessing. But with smaller dataset multiprocessing optimisation takes more time than without it.
- Find a way to resampling all of the matrices faster : indeed, the bigger the dataset is, the longer it will take. In the one that I'm using, resampling all of the matrices takes more than 52 seconds. An issue should be make to adressed this specific problem.

Aurelienpujolm committed:
Syntax of comment

Co-authored-by: Alex Dastous <47249340+po09i@users.noreply.github.com>

Aurelienpujolm committed:
Fixed a mispell in a comment

Co-authored-by: Alex Dastous <47249340+po09i@users.noreply.github.com>

Aurelienpujolm committed:
Fixed syntax in a comment

Co-authored-by: Alex Dastous <47249340+po09i@users.noreply.github.com>

Aurelienpujolm committed:
Fixed syntax in a comment

Co-authored-by: Alex Dastous <47249340+po09i@users.noreply.github.com>

Aurelienpujolm committed:
Apply suggestions from code review

Co-authored-by: Alex Dastous <47249340+po09i@users.noreply.github.com>

Aurelienpujolm committed:
Apply suggestions from code review

Co-authored-by: Alex Dastous <47249340+po09i@users.noreply.github.com>

Aurelienpujolm committed:
Fixed a problem with the reviews changes

Aurelienpujolm committed:
Apply the review changes
Put a continue in the for loop of eval_static_shim
Apply the correct indentation

Aurelienpujolm committed:
Fixed the bug in plot_coefs + changes in calc_shimmed_anat_orient
make sure that the program doesn't crash if the slices dimension is different than 1

Aurelienpujolm committed:
Fixed the cases where all the slices are shim in plot_coefs

Aurelienpujolm committed:
Put Multiprocessing for every dataset (May change)

po09i commented on shimmingtoolbox/cli/b0shim.py:
```diff
@@ -1018,32 +1033,58 @@ def _plot_coefs(coil, slices, static_coefs, path_output, coil_number, rt_coefs=N
-        # Draw a black line at y=0
-        ax.hlines(0, 0, 1, transform=ax.get_yaxis_transform(), colors='k')
+def _add_sub_figure(i_shim, fig, n_shims, axis, rt_coefs, pres_probe_min, pres_probe_max, bounds, min_y, max_y,
+                    units, static_coefs, i_slice):
```
n_shims seems to be called with the same number whether there are unused slices are not. (number_slices_shimmed)

Aurelienpujolm replied:
Yes, You would prefer to redeclare nshims, everytime that we enter the function? I thaught it would be better to do it in plot_coefs because we get this variable with slice in plot_coefs, and so it is better to only calculate it once for me.

po09i replied:
I meant whether there are unused slices or not, n_shims is still the same number which should not occur since n)shims describes the number of subplots. You are not seeing an error because you added 1 and are not using the last subplot if there are no unshimmed slice. I fixed it.

po09i committed:
Refactor _plot_coefs

po09i commented on shimmingtoolbox/optimizer/lsq_optimizer.py:
```diff
@@ -208,6 +216,32 @@ def _initial_guess_zeros(self):
+        np.mean((unshimmed_vec + np.sum(coil_mat * coef, axis=1, keepdims=False)) ** 2) / factor+\
+           (self.reg_factor * np.mean(np.abs(coef) / self.reg_factor_channel))
+        If you want to see the mathematical expression for the jacobian, of this function or for the mae one,
+        please contact : aurelienpujolmpro@gmail.com
```
Write it on GitHub so we don't need to email you in the future please.

jcohenadad replied:
👉  https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/writing-mathematical-expressions

po09i committed:
Refactor optimizer

po09i committed:
Small fixups

po09i committed:
Comment mp bug that outputs debug even when not in debug

po09i committed:
Move plotting of currents after saving the coefficients

po09i committed:
Remove newline
